rust-runtime/smithy-query/src/lib.rs

```diff
@@ -0,0 +1,388 @@
+            self.prefix,
+            entry,
+            self.next_index,
+            encode(key)
+        ));
+        let value_name = format!("{}{}.{}.value", self.prefix, entry, self.next_index);
+
+        self.next_index += 1;
+        QueryValueWriter::new(self.output, Cow::Owned(value_name))
+    }
+
+    pub fn finish(self) {
+        // Calling this drops self
+    }
+}
+
+pub struct QueryListWriter<'a> {
+    output: &'a mut String,
+    prefix: Cow<'a, str>,
+    flatten: bool,
+    member_override: Option<&'a str>,
+    next_index: usize,
+}
+
+impl<'a> QueryListWriter<'a> {
+    fn new(
+        output: &'a mut String,
+        prefix: Cow<'a, str>,
+        flatten: bool,
+        member_override: Option<&'a str>,
+    ) -> QueryListWriter<'a> {
+        QueryListWriter {
+            prefix,
+            output,
+            flatten,
+            member_override,
+            next_index: 1,
+        }
+    }
+
+    pub fn entry(&mut self) -> QueryValueWriter {
+        let value_name = if self.flatten {
+            format!("{}.{}", self.prefix, self.next_index)
+        } else if self.member_override.is_some() {
+            format!(
+                "{}.{}.{}",
+                self.prefix,
+                self.member_override.unwrap(),
+                self.next_index
+            )
+        } else {
+            format!("{}.member.{}", self.prefix, self.next_index)
+        };
+
+        self.next_index += 1;
+        QueryValueWriter::new(self.output, Cow::Owned(value_name))
+    }
+
+    pub fn finish(self) {
+        // Calling this drops self
+    }
+}
+
+pub struct QueryValueWriter<'a> {
+    output: &'a mut String,
+    prefix: Cow<'a, str>,
+}
+
+impl<'a> QueryValueWriter<'a> {
+    pub fn new(output: &'a mut String, prefix: Cow<'a, str>) -> QueryValueWriter<'a> {
+        QueryValueWriter { output, prefix }
+    }
+
+    /// Starts a new prefix.
+    pub fn prefix(&mut self, prefix: &'a str) -> QueryValueWriter {
+        QueryValueWriter::new(
+            self.output,
+            Cow::Owned(format!("{}.{}", self.prefix, prefix)),
+        )
+    }
+
+    /// Writes the boolean `value`.
+    pub fn boolean(mut self, value: bool) {
+        self.write_param_name();
+        self.output.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+    }
+
+    /// Writes a string `value`.
+    pub fn string(mut self, value: &str) {
+        self.write_param_name();
+        self.output.push_str(&encode(value));
+    }
+
+    /// Writes a number `value`.
+    pub fn number(self, value: Number) {
+        match value {
+            Number::PosInt(value) => {
+                // itoa::Buffer is a fixed-size stack allocation, so this is cheap
+                self.string(itoa::Buffer::new().format(value));
+            }
+            Number::NegInt(value) => {
+                self.string(itoa::Buffer::new().format(value));
+            }
+            Number::Float(value) => {
```

Comment:
Does aws query have these float semantics?
I'm not positive we want to go through the number interface here. In other places we just directly format (or call iron on in the type, whatever numeric type it happens to be)

Reply:
The floating point behavior isn't specified for NaN of infinite values. I opened an issue on Smithy to probe into this: https://github.com/awslabs/smithy/issues/812

My stance right now is to keep it consistent with the JSON implementation so there are no surprises. We may even want to make XML consistent in the meantime.